File: Cpp/Arrays/Prob3_BoyerMooreMajorityAlgorithm.cpp
/**
 * @file Prob3_BoyerMooreMajorityAlgorithm.cpp
 * @author your name (you@domain.com)
 * @brief
 * @version 0.1
 * @date 2023-10-15
 *
 * @copyright Copyright (c) 2023
 *
 * Boyer-Moore Majority Voting Algorithm
 */

#include <iostream>

using namespace std;

int findMajorityNumber(int *arr, int size)
{

    int votes = 0;
    int candidate = -1;

    for (int i = 0; i < size; i++)
    {
        if (votes > 0)
        {
            if (candidate == arr[i])
            {
                votes++;
            }
            else
            {
                votes--;
            }
        }
        else
        {
            candidate = arr[i];
            votes++;
        }
    }

    int count = 0;

    for (int i = 0; i < size; i++)
    {
        if (arr[i] == candidate)
        {
            count++;
        }
    }

    if (count > size / 2)
    {
        return candidate;
    }

    return -1;
}

int main()
{

    int arr[] = {1, 1, 1, 1, 1, 2, 3, 4, 5};
    int size = sizeof(arr) / sizeof(arr[0]);
    cout << "Majority Number: " << findMajorityNumber(arr, size);
    return 0;
}

File: Python/Arrays/Prob3_BoyerMooreMajorityAlgorithm.py
# Boyer-Moore Majority Voting Algorithm

def findMajorityNumber(arr):
    l = len(arr)
    votes = 0
    candidate = -1

    for i in range(l):
        if votes == 0:
            candidate = arr[i]
            votes += 1
        else:
            if arr[i] == candidate:
                votes += 1
            else:
                votes -= 1

    count = 0
    for i in range(l):
        if (arr[i] == candidate):
            count += 1

    return candidate if count > l // 2 else -1


if __name__ == "__main__":
    arr = [1, 1, 2, 2, 2, 2, 2, 4, 5]
    print("Majority number:", findMajorityNumber(arr))
